### ISSUE TITLE
Support no launch.json debugging

### ISSUE THREAD

**Issue by isidorn** [jest-community/vscode-jest#776]
Hi VS Code PM here 👋 

We would love if Jest debug would support debugging without a `launch.json`. Thus allowing users to debug in a no folder vscode.
Easiest to do this is to have a dynamic `resolveDebugConfiguartion` that will provide a launch configuration on the fly. [Example](https://github.com/microsoft/vscode-mock-debug/blob/main/src/activateMockDebug.ts#L161)

Apart from that consider contributing to the Editor title area menu, such that a Play button would appear in the Editor Title area (picture below). Example how to do this https://github.com/microsoft/vscode-mock-debug/blob/main/package.json#L80

I think both of these are super important and would really love that we improve this experience 💯  
Let me know if I can help somehow. I am also open to a quick 30 min meeting where we can discuss about the new cool VS Code debug features that Jest debug could potentially adopt.

**Comment by connectdotz** [comment]
hi @isidorn this is interesting... in a way we did support "debugging without a launch.json" scenario via our own `DebugConfigurationProvider`, which generates a default debug-config if none is available upon debugging. We are happy to support new API, but a few questions:

- in `resolveDebugConfiguartion`, where is the debug config coming from? Since we are looking for a specific type of debug-config, not sure how it could be handled in this situation.
- I'm not familiar with how the "folder-less" system works, just wondering how will a user customize their debug config in that scenario?
- we actually were thinking to replace our debug codeLens with TestExplorer's debug menu going forward, so will be interested to see how TestExplorer plan to support this feature and will it require test API changes... ( /cc @connor4312 )
- regarding the editor title menu, is there a plan to put the TestExplorer's menu up there?  If the answer is yes then we don't need to be redundant. ( /cc @connor4312 )

**Comment by connor4312** [comment]
> regarding the editor title menu, is there a plan to put the TestExplorer's menu up there? If the answer is yes then we don't need to be redundant

Most likely, yes

**Comment by isidorn** [comment]
@connectdotz thanks for your response. 
Great that you plan to jump on the TestExplorer, though @connor4312 has more context there.

1. It is up to the extension to decide how to compute the debug config in the `resolveDebugConfiguration`. Is your extension a bit special that it relies on the js-debug? I am not familiar with `jest` I have to confess.
2. Imagine a folder less system as being something for setting up and trying out things. There is no way for a user to customise their debug config. We just want something that would work for the majority of cases, such that users can simply run / debug single files

**Comment by connectdotz** [comment]
>  Is your extension a bit special that it relies on the js-debug?

Yes indeed, we modify/generate debug config then pass that to the js debugger (`vscode.debug.startDebuggin()`)

> Imagine a folder less system as being something for setting up and trying out things. There is no way for a user to customise their debug config. We just want something that would work for the majority of cases, such that users can simply run / debug single files

If that is the main use case then I think people should be able to use our DebugCodeLen or TestExplorer debug function just fine.

Please let us know if there is a way for us to verify this, thanks.

**Comment by isidorn** [comment]
@connectdotz the easiest to verify is to just Open VS Code File > Close Folder
Now you are in a Purple Status Bar No Folder VS Code. Yey. Now open a jest file. Try to run it. If it works all is good :)

**Comment by connectdotz** [comment]
@isidorn sadly, that did not work... and it is beyond debug, the extension did not get fully activated if `vscode.workspace.workspaceFolders` is undefined. There will be no debug codeLens nor TestExplorer for this kind of use cases. The workspace is a pretty fundamental piece in our extension, I am afraid it will require a major change to support  no-workspace use case...

**Comment by isidorn** [comment]
Thanks for trying. If the workspace is a fundamental piece of your extension I understand how this would not be possible for now. Though maybe you can consider it in the future? Anyways feel free to close this item if you do not think it is in reach. Thanks!

**Comment by connectdotz** [comment]
@isidorn I will close this ticket for now. But yes, we could reopen it if there is high demand for supporting this use case in the future...
